### ISSUE TITLE
`jax.numpy.ndarray.at[].set()` much slower than numpy on GPU?

### ISSUE THREAD

**Issue by epignatelli** [google/jax#16587]
I am sorry -- I have been opening new issues quite frequently on indexing in the last couple of days and that might me annoying. I am just finding some limitations in the indexing operations, and, for each I wonder if there's is a solution (or if it is a fake problem).

This might be a microbenchmark, but when a set of functions rely heavily on this mechanism, updates becomes an issue.
I see this is also the case for TF https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update, so perhaps not a limitation of jax per sé.

```python
import jax
import jax.numpy as jnp
import numpy as np
from timeit import repeat
import matplotlib.pyplot as plt

            
def set_jax(a, idx):
    return a.at[idx].set(idx)

def set_numpy(a, idx):
    a[idx] = idx
    return a

results = []
for i in range(1, 8):
    shape = (10 ** i,)
    idx = i

    # jax cpu
    cpu = jax.devices("cpu")[0]
    jax_array_cpu = jax.random.uniform(jax.random.PRNGKey(0), shape=shape)
    jax_array_cpu = jax.device_put(jax_array_cpu, cpu)
    idx_jax_cpu = jax.device_put(jnp.asarray(idx), cpu)
    
    set_jax_cpu = jax.jit(set_jax, backend="cpu").lower(jax_array_cpu, idx_jax_cpu).compile()
    jax_cpu_res = repeat(lambda: set_jax_cpu(jax_array_cpu, idx_jax_cpu).block_until_ready(), number=10, repeat=100)

    # jax gpu
    gpu = jax.devices("gpu")[0]
    jax_array_gpu =  jax.random.uniform(jax.random.PRNGKey(0), shape=shape)
    jax_array_gpu = jax.device_put(jax_array_gpu, gpu)
    idx_jax_gpu = jax.device_put(jnp.asarray(idx), gpu)
    
    set_jax_gpu = jax.jit(set_jax, backend="gpu").lower(jax_array_gpu, idx_jax_gpu).compile()
    jax_gpu_res = repeat(lambda: set_jax_gpu(jax_array_gpu, idx_jax_gpu).block_until_ready(), number=10, repeat=100)

    # np
    numpy_array = np.random.uniform(size=shape)
    numpy_res = repeat(lambda: set_numpy(numpy_array, idx), number=10, repeat=100)

    # results
    jax_cpu_res = (sum(jax_cpu_res) / len(jax_cpu_res))
    jax_gpu_res = (sum(jax_gpu_res) / len(jax_gpu_res))
    numpy_res = (sum(numpy_res) / len(numpy_res))
    results.append((jax_cpu_res, jax_gpu_res, numpy_res))

print(jnp.asarray(results))
jax_cpu_res, jax_gpu_res, numpy_res = list(zip(*results))
plt.plot(jax_cpu_res)
plt.plot(jax_gpu_res)
plt.plot(numpy_res)
plt.yscale("log")
plt.legend(["jax (cpu)", "jax (gpu)", "numpy"])
plt.show()
```

![image](https://github.com/google/jax/assets/26899347/d89a6a79-1ffb-418b-948d-2b55c38c2270)

The x axis is shifted by `-1`, so `0` should read`1`. 

Also, I am sure I am misinterpreting because I don't know XLA, but it seems like the indexed array is copied at every update (`  %copy.1 = f32[100]{0} copy(f32[100]{0} %Arg_0.1)`), which is the opposite of what the [docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html#jax.numpy.ndarray.at:~:text=However%2C%20inside%20a%20jit()%20compiled%20function%2C%20expressions%20like%20x%20%3D%20x.at%5Bidx%5D.set(y)%20are%20guaranteed%20to%20be%20applied%20in%2Dplace.) say?

```python
print(jax.jit(set_jax, backend="gpu").lower(jax_array_gpu, idx_jax_gpu).compile().as_text())
```

```
HloModule jit_set_jax, is_scheduled=true, entry_computation_layout={(f32[100]{0}, s32[])->f32[100]{0}}

%fused_computation (param_0: f32[100], param_1.2: s32[]) -> f32[100] {
  %param_0 = f32[100]{0} parameter(0)
  %constant_40 = s32[1]{0} constant({0})
  %param_1.2 = s32[] parameter(1)
  %constant_39 = s32[] constant(0)
  %compare.12 = pred[] compare(s32[] %param_1.2, s32[] %constant_39), direction=LT, metadata={op_name="jit(set_jax)/jit(main)/lt" source_file="<ipython-input-14-b6c77a8357d6>" source_line=9}
  %constant_38 = s32[] constant(100)
  %add.4 = s32[] add(s32[] %param_1.2, s32[] %constant_38), metadata={op_name="jit(set_jax)/jit(main)/add" source_file="<ipython-input-14-b6c77a8357d6>" source_line=9}
  %select.6 = s32[] select(pred[] %compare.12, s32[] %add.4, s32[] %param_1.2), metadata={op_name="jit(set_jax)/jit(main)/select_n" source_file="<ipython-input-14-b6c77a8357d6>" source_line=9}
  %bitcast.62 = s32[1]{0} bitcast(s32[] %select.6)
  %compare.11 = pred[1]{0} compare(s32[1]{0} %constant_40, s32[1]{0} %bitcast.62), direction=LE
  %constant_37 = s32[1]{0} constant({99})
  %compare.10 = pred[1]{0} compare(s32[1]{0} %constant_37, s32[1]{0} %bitcast.62), direction=GE
  %and.4 = pred[1]{0} and(pred[1]{0} %compare.11, pred[1]{0} %compare.10)
  %convert.1 = f32[] convert(s32[] %param_1.2), metadata={op_name="jit(set_jax)/jit(main)/convert_element_type[new_dtype=float32 weak_type=False]" source_file="<ipython-input-14-b6c77a8357d6>" source_line=9}
  %bitcast.61 = f32[1]{0} bitcast(f32[] %convert.1)
  %dynamic-slice.5 = f32[1]{0} dynamic-slice(f32[100]{0} %param_0, s32[] %select.6), dynamic_slice_sizes={1}
  %select.5 = f32[1]{0} select(pred[1]{0} %and.4, f32[1]{0} %bitcast.61, f32[1]{0} %dynamic-slice.5)
  ROOT %dynamic-update-slice.3 = f32[100]{0} dynamic-update-slice(f32[100]{0} %param_0, f32[1]{0} %select.5, s32[] %select.6), sharding={replicated}, metadata={op_name="jit(set_jax)/jit(main)/scatter[update_consts=() dimension_numbers=ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,)) indices_are_sorted=True unique_indices=True mode=GatherScatterMode.FILL_OR_DROP]" source_file="<ipython-input-14-b6c77a8357d6>" source_line=9}
}

ENTRY %main.16 (Arg_0.1: f32[100], Arg_1.2: s32[]) -> f32[100] {
  %Arg_1.2 = s32[] parameter(1), sharding={replicated}
  %Arg_0.1 = f32[100]{0} parameter(0), sharding={replicated}
  %copy.1 = f32[100]{0} copy(f32[100]{0} %Arg_0.1)
  ROOT %fusion = f32[100]{0} fusion(f32[100]{0} %copy.1, s32[] %Arg_1.2), kind=kLoop, calls=%fused_computation, sharding={replicated}, frontend_attributes={fingerprint_before_lhs="08956b1478b109d4ebe2d6dc5d47afc7"}, metadata={op_name="jit(set_jax)/jit(main)/scatter[update_consts=() dimension_numbers=ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,)) indices_are_sorted=True unique_indices=True mode=GatherScatterMode.FILL_OR_DROP]" source_file="<ipython-input-14-b6c77a8357d6>" source_line=9}
}
```

Potentially related to @patrick-kidger's [comment](https://github.com/google/jax/issues/16559#issuecomment-1610019539).

**Comment by jakevdp** [comment]
Hi - this is expected, because in the JAX CPU and GPU  versions, you're not only benchmarking array updates, but array copies. The reason is that, in the Python world, no two JAX arrays can share the same memory. So when you do a micro-benchmark of a numpy function that returns a view vs. a JAX function that returns a copy, the relevant difference will be the cost of the copy.

In the course of a real program, this is not an issue, because you generally will not be returning the result of the array update in a Python function, and so no copy needs to be made.

For more on what to expect when comparing microbenchmarks of JAX and NumPy code, see [FAQ: Is JAX Faster than NumPy](https://jax.readthedocs.io/en/latest/faq.html#is-jax-faster-than-numpy).

**Comment by jakevdp** [comment]
If you want a fairer comparison of JAX and NumPy, you could do something like this:
```python
def set_jax(a, idx):
    return a.at[idx].set(idx).sum()

def set_numpy(a, idx):
    a[idx] = idx
    return a.sum()
```
Returning `a.sum()` makes the JAX and numpy code more-or-less semantically equivalent, and you can see that jax-on-cpu is similar to numpy, but jax-on-gpu beats numpy (as one would expect). The main computation cost here is the `sum`, because the updates are in-place and virtually free in all three cases:
![download](https://github.com/google/jax/assets/781659/6cb70245-5076-4df9-beb3-a50235783fbb)

**Comment by epignatelli** [comment]
> In short, it's not the update that's causing your array to be copied, it's your return statement.

Golden info, thank you very much for the thorough answer, Jake.
